src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking for function contracts
+
+Author: Remi Delmas, delmasrd@amazon.com
+Date: November 2022
+
+\*******************************************************************/
+
+/// \file
+/// Collects all user-defined predicates that call functions
+/// is_fresh, pointer_in_range, obeys_contract and lifts them to function taking
+/// pointers by reference. When called in assumption contexts,
+/// These user-defined predicates update the pointers using side effect in order
+/// to make the assumptions described by the predicate hold.
+/// In assertion contexts, they behave like the original user-defined predicate
+/// i.e. without side effects.
+
+#ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_LIFT_MEMORY_PREDICATES_H
+#define CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_LIFT_MEMORY_PREDICATES_H
+
+#include <util/message.h>
+
+#include <goto-programs/goto_program.h>
+
+#include <map>
+#include <set>
+
+class dfcc_utilst;
+class dfcc_libraryt;
+class dfcc_instrumentt;
+class message_handlert;
+class goto_modelt;
+class exprt;
+class replace_symbolt;
+
+class dfcc_lift_memory_predicatest
+{
+public:
+  /// \param goto_model The goto model to process
+  /// \param utils Utility methods
+  /// \param library The contracts instrumentation library
+  /// \param instrument The DFCC instrumenter object
+  /// \param message_handler Used for messages
+  dfcc_lift_memory_predicatest(
+    goto_modelt &goto_model,
+    dfcc_utilst &utils,
+    dfcc_libraryt &library,
+    dfcc_instrumentt &instrument,
+    message_handlert &message_handler);
+
+  /// \brief Collects and lifts all user-defined memory predicates.
+  /// \param[out] lifted Set of predicates that were lifted
+  /// \param[out] discovered_function_pointer_contracts Set of function pointer
+  /// contracts discovered during instrumentation
+  void lift_predicates(
+    std::set<irep_idt> &lifted,
+    std::set<irep_idt> &discovered_function_pointer_contracts);
+
+  /// Fixes calls to user-defined memory predicates in the given program,
+  /// by adding an address-of to arguments passed in lifted position.
+  void fix_calls(goto_programt &program);
+
+  /// Fixes calls to user-defined memory predicates in the given program,
+  /// by adding an address-of to arguments passed in lifted position, between
+  /// first_instruction (included) and last_instruction (excluded).
+  void fix_calls(
+    goto_programt &program,
+    goto_programt::targett first_instruction,
+    const goto_programt::targett &last_instruction);
+
+protected:
+  goto_modelt &goto_model;
+  dfcc_utilst &utils;
+  dfcc_libraryt &library;
+  dfcc_instrumentt &instrument;
+  message_handlert &message_handler;
+  messaget log;
```

Comment:
Do you actually need both of these? You could always get the message handler from `log` using `get_message_handler()`.

Reply:
Fixed